File: rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/fus/UnityProjectUsageCollector.kt
@file:Suppress("UnstableApiUsage")

package com.jetbrains.rider.plugins.unity.fus

import com.intellij.internal.statistic.beans.MetricEvent
import com.intellij.internal.statistic.eventLog.EventLogGroup
import com.intellij.internal.statistic.eventLog.events.EventFields
import com.intellij.internal.statistic.service.fus.collectors.ProjectUsagesCollector
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.intellij.util.ui.UIUtil
import com.jetbrains.rd.platform.util.createNestedAsyncPromise
import com.jetbrains.rd.platform.util.lifetime
import com.jetbrains.rd.util.lifetime.isNotAlive
import com.jetbrains.rd.util.lifetime.onTermination
import com.jetbrains.rider.plugins.unity.model.frontendBackend.frontendBackendModel
import com.jetbrains.rider.projectView.solution
import org.jetbrains.concurrency.CancellablePromise

class UnityProjectUsageCollector  : ProjectUsagesCollector() {
    companion object {
        private val RIDER_UNITY_STATE_GROUP = EventLogGroup("rider.unity.state", 1)

        private val UnityTechnology = RIDER_UNITY_STATE_GROUP.registerEvent(
            "unityTechnology",
            EventFields.String("id", mutableListOf("HDRP", "CoreRP", "URP", "ECS", "InputSystem", "Burst", "Odin", "Peek", "UniRx", "UniTask")),
            EventFields.Boolean("isDiscovered"),
        )
    }

    override fun getGroup(): EventLogGroup {
        return RIDER_UNITY_STATE_GROUP
    }

    override fun getMetrics(project: Project, indicator: ProgressIndicator?): CancellablePromise<out MutableSet<MetricEvent>> {
        val promise = project.lifetime.createNestedAsyncPromise<MutableSet<MetricEvent>>()

        UIUtil.invokeLaterIfNeeded {
            val result = mutableSetOf<MetricEvent>()
            if (project.lifetime.isNotAlive) {
                return@invokeLaterIfNeeded
            }
            val lifetimeDef = project.lifetime.createNested()
            val model = project.solution.frontendBackendModel
            model.isTechnologyDiscoveringFinished.advise(lifetimeDef.lifetime) {
                if (it) {
                    for ((id, state) in model.discoveredTechnologies) {
                        result.add(UnityTechnology.metric(id, state))
                    }
                    promise.setResult(result)
                    lifetimeDef.terminate()
                }
            }
        }
        return promise
    }
}

File: resharper/resharper-unity/src/Unity.Rider/Integration/Core/Feature/Services/FeatureStatistics/UnityProjectUsageCollector.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using JetBrains.Collections;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;
using JetBrains.ProjectModel;
using JetBrains.ReSharper.Feature.Services.FeaturesStatistics;
using JetBrains.ReSharper.Plugins.Unity.Core.ProjectModel;
using JetBrains.ReSharper.Resources.Shell;
using JetBrains.UsageStatistics.FUS.EventLog;
using JetBrains.UsageStatistics.FUS.EventLog.Events;
using JetBrains.UsageStatistics.FUS.EventLog.Fus;
using JetBrains.Util;

namespace JetBrains.ReSharper.Plugins.Unity.Rider.Integration.Core.Feature.Services.FeatureStatistics
{
    [SolutionComponent]
    public class UnityProjectUsageCollector : SolutionUsagesCollector
    {
        private readonly UnitySolutionTracker myUnitySolutionTracker;
        private readonly UnityTechnologyDescriptionCollector myDescriptionCollector;
        private EventLogGroup myGroup;
        private readonly EventId2<string, bool> myEvent;

        public UnityProjectUsageCollector(UnitySolutionTracker unitySolutionTracker, UnityTechnologyDescriptionCollector descriptionCollector, FeatureUsageLogger featureUsageLogger)
        {
            myUnitySolutionTracker = unitySolutionTracker;
            myDescriptionCollector = descriptionCollector;
            myGroup = new EventLogGroup("dotnet.unity.state", "Packages and Assets used by Unity users", 3, featureUsageLogger);
            myEvent = myGroup.RegisterEvent("unityTechnology", "Unity Technology", EventFields.String("id", "Name",
                new[] { "HDRP", "CoreRP", "URP", "ECS", "InputSystem", "Burst", "Odin", "Photon", "Peek", "UniRx",
                    "UniTask", 
                    "TestFramework",
                    "PythonScripting",
                    "Addressables",
                    "AndroidLogCat",
                    "CodeCoverage",
                    "UnityCollections",
                    "EditorCoroutines",
                    "EntitiesGraphics",
                    "Localization",
                    "Mathematics",
                    "Transport",
                    "UnityPhysics",
                    "Havok",
                    "MlAgents",
                    "MultiplayerTools",
                    "NetCode",
                    "NetCodeGameObjects",
                    "Serialization",
                    "Logging",
                    "MemoryProfiler",
                    "ProfilerAnalyzer",
                    "ProfilerCore",
                    "Collab",
                    "VisualScripting"
                }), EventFields.Boolean("isDiscovered", "Is Discovered"));
        }
        
        public override EventLogGroup GetGroup()
        {
            return myGroup;
        }


        public override async Task<ISet<MetricEvent>> GetMetricsAsync(Lifetime lifetime)
        {
            if (!myUnitySolutionTracker.IsUnityProject.HasTrueValue())
                return EmptySet<MetricEvent>.Instance;
            
            return await lifetime.StartMainReadAsync(async () =>
            {
                await myDescriptionCollector.Ready.NextTrueValueAsync(lifetime);
                var hashSet = new HashSet<MetricEvent>();
                foreach (var (key, value) in myDescriptionCollector.DiscoveredTechnologies)
                {
                    hashSet.Add(myEvent.Metric(key, value));
                }

                return hashSet;
            });
        }
    }
}